lib_nn/api/AbstractKernel.hpp

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include "geom/Filter2dGeometry.hpp"
+
+namespace nn {
+
+/**
+ * Class representing an executable filter kernel. A concrete instance of this
+ * class processes a particular region of the output image associated with the
+ * filter.
+ */
+class AbstractKernel {
+ public:
+  /**
+   * Struct indicating a (3D) rectangular sub-region of an image, as well as how
+   * to iterate through pixels within that region of the image.
+   *
+   * @see AbstractKernel
+   */
+  class Params {
+   public:
+    /**
+     * The first (`h_begin`; inclusive) and final (`h_end`; exclusive) rows of
+     * the output image to be processed when the corresponding filter is
+     * executed.
+     */
+    const int32_t h_begin, h_end;
+
+    /**
+     * The first (`w_begin`; inclusive) and final (`w_end`; exclusive) columns
+     * of the output image to be processed when the corresponding filter is
+     * executed.
+     */
+    const int32_t w_begin, w_end;
+
+    /**
+     * The number of output channel groups that will be processed when this
+     * filter is executed.
+     */
+    int32_t output_channel_group_count;
+
+    // Used for setting the first channels slice, i.e. rather than writing to
+    // slice 0-31 by offsetting it we can address slice 32 - 63, etc.
+
+    /**
+     * The first output channel to be processed when this filter is executed.
```

Comment:
Is this comment discussing something else?

Reply:
no, I'll merge them together